pkg/vmprovider/providers/vsphere/session/session_vm_customization.go

```diff
@@ -245,22 +278,33 @@ func customizeCloudInit(
 
 	cloudInitMetadata, err := GetCloudInitMetadata(vmCtx.VM, netplan, updateArgs.VMMetadata.Data)
 	if err != nil {
+		conditions.MarkFalse(vmCtx.VM,
+			vmopv1.VirtualMachineMetadataReadyCondition,
+			vmopv1.VirtualMachineMetadataInvalidReason,
+			vmopv1.ConditionSeverityError,
+			err.Error())
 		return nil, nil, err
 	}
 
 	var configSpec *vimTypes.VirtualMachineConfigSpec
 	var custSpec *vimTypes.CustomizationSpec
+	var errWithReason *ErrWithReason
 
 	switch vmCtx.VM.Annotations[constants.CloudInitTypeAnnotation] {
 	case constants.CloudInitTypeValueCloudInitPrep:
-		configSpec, custSpec, err = GetCloudInitPrepCustSpec(cloudInitMetadata, updateArgs)
+		configSpec, custSpec, errWithReason = GetCloudInitPrepCustSpec(cloudInitMetadata, updateArgs)
 	case constants.CloudInitTypeValueGuestInfo, "":
 		fallthrough
 	default:
-		configSpec, err = GetCloudInitGuestInfoCustSpec(cloudInitMetadata, config, updateArgs)
+		configSpec, errWithReason = GetCloudInitGuestInfoCustSpec(cloudInitMetadata, config, updateArgs)
 	}
 
-	if err != nil {
+	if errWithReason != nil {
+		conditions.MarkFalse(vmCtx.VM,
+			vmopv1.VirtualMachineMetadataReadyCondition,
+			errWithReason.conditionReason,
+			vmopv1.ConditionSeverityError,
+			errWithReason.Error())
 		return nil, nil, err
```

Comment:
This is no longer the correct `err`. The pattern we're introducing here - an error wrapper that doesn't implement Error() - makes me concerned that similar bugs will continue to crop up.

In v1a2 - BTW please add the appropriate types to v1a2 too - we have have finer grain conditions, and I've gone a little back and forward with a similar approach here and setting the conditions closer to were the error occurs. I settled the later for now; in various other projects I've seen a `func errWithCondition(...) error` so the usage becomes like `return nil, errWithCondition(obj, MyCondition, "Reason", err)` but I haven't had the time to really think about the ergonomics. I do want us to settle on a consistent pattern of whose responsibility it is to set the condition.

Reply:
Since the new type `ErrWithReason` contains an embedded `error` type, the function call `customErr.Error()` is valid and would work. [Here](https://go.dev/play/p/CB6qPRQk1tD) is a golfing playground snippet I tried out to confirm this.

💯 To have a consistent behavior for all conditions, I think we should rally behind a consistent pattern when making the move onto the v1a2 types.

Reply:
Oh, yes, I missed that error was embedded this morning. My comment about the usage is still applicable though: the code is `return nil, nil, err`, which `err` will now always be `nil` at this point because the change introduced `errWithReason`. Having multiple err-vars in scope is too likely for the wrong one to be returned that we should really avoid that.